File: packages/agent-core/src/tools/builtin/planning/exit-plan-mode.ts
/**
 * ExitPlanModeTool — plan-mode exit tool.
 *
 * The LLM calls this tool to surface a finalised plan to the user and
 * exit plan mode. The plan must already be written to the current plan
 * file; this tool reads that file and flips plan mode off.
 */

import type { Agent } from '#agent';
import type { OdyConfig } from '@odysseythink/agent-core-shared';
import type { SessionModeData } from '#agent/session-mode';
import type { Kaos } from '@odysseythink/kaos';
import { z } from 'zod';

import type { BuiltinTool } from '../../../agent/tool';
import type { ExecutableToolResult, ToolExecution } from '../../../loop/types';
import type { ToolInputDisplay } from '../../display';
import { toInputJsonSchema } from '../../support/input-schema';
import {
  E2EPlanEnricher,
  E2EConfigResolver,
  registry,
} from '@odysseythink/e2e-testing';
import {
  declaredOptionLabel,
  isViaApproval,
  selectedApproachPrefix,
  selectedLabelOf,
} from './exit-mode-output';
import DESCRIPTION from './exit-plan-mode.md';

// ── Input schema ─────────────────────────────────────────────────────

/**
 * User-selectable option surfaced at plan approval time. The LLM supplies
 * up to 3 of these when the plan contains multiple approaches; the host's
 * ApprovalRuntime presents them to the user and returns the chosen `label`
 * (or `{kind:'revise', feedback}` when the user asks for revisions).
 */
export interface ExitPlanModeOption {
  label: string;
  description: string;
}

export interface ExitPlanModeInput {
  options?: readonly ExitPlanModeOption[] | undefined;
}

const RESERVED_OPTION_LABELS = new Set(
  ['Approve', 'Reject', 'Reject and Exit', 'Revise'].map(normalizeOptionLabel),
);

const ExitPlanModeOptionSchema = z
  .object({
    label: z
      .string()
      .min(1)
      .max(80)
      .describe(
        'Short name for this option (1-8 words). Append "(Recommended)" if you recommend this option.',
      ),
    description: z
      .string()
      .default('')
      .describe('Brief summary of this approach and its trade-offs.'),
  })
  .strict();

export const ExitPlanModeInputSchema: z.ZodType<ExitPlanModeInput> = z
  .object({
    options: z
      .array(ExitPlanModeOptionSchema)
      .min(1)
      .max(3)
      .refine(hasUniqueOptionLabels, 'Option labels must be unique.')
      .refine(hasNoReservedOptionLabels, 'Option labels must not use reserved approval labels.')
      .optional()
      .describe(
        'When the plan contains multiple alternative approaches, list them here so the user can choose which one to execute. Provide up to 3 options; 2-3 distinct approaches work best when the plan offers a real choice. Passing a single option is allowed and is equivalent to a plain plan approval. Each option represents a distinct approach from the plan. Do not use "Reject", "Revise", "Approve", or "Reject and Exit" as labels.',
      ),
  })
  .strict();

export interface ExitPlanModePlanSource {
  plan: string;
  path?: string | undefined;
}

type ResolvePlanResult =
  | { readonly ok: true; readonly plan: string; readonly path?: string | undefined }
  | { readonly ok: false; readonly error: ExecutableToolResult };

// ── Implementation ───────────────────────────────────────────────────

export class ExitPlanModeTool implements BuiltinTool<ExitPlanModeInput> {
  readonly name = 'ExitPlanMode' as const;
  readonly description: string = DESCRIPTION;
  readonly parameters: Record<string, unknown> = toInputJsonSchema(ExitPlanModeInputSchema);

  constructor(private readonly agent: Agent, private readonly kaos: Kaos) {}

  async resolveExecution(args: ExitPlanModeInput): Promise<ToolExecution> {
    return {
      description: 'Presenting plan and exiting plan mode',
      display: await this.resolvePlanReviewDisplay(args),
      approvalRule: this.name,
      execute: (ctx) => this.execution(args, ctx.metadata),
    };
  }

  private async maybeEnrichPlanForE2E(): Promise<void> {
    try {
      const e2eConfig = E2EConfigResolver.resolve(this.agent.kimiConfig ?? ({} as OdyConfig));
      if (!e2eConfig.enabled) return;

      const modeData = await this.agent.sessionMode.data();
      if (modeData === null || modeData.kind !== 'plan') return;
      if (modeData.content.trim().length === 0 || modeData.path.length === 0) return;

      // Use the project's own generator (Go, Python, TS/Vitest, …) to decide
      // which features are affected, so user projects — not just ody-code's
      // builtin tools — get an E2E task injected. No matching generator means
      // we cannot generate tests for this project, so skip enrichment.
      const projectRoot = this.kaos.getcwd();
      let generator;
      try {
        generator = await registry.detectAndGet(projectRoot);
      } catch {
        return;
      }

      const enricher = new E2EPlanEnricher(this.kaos, e2eConfig, {
        analyze: (files, config) => generator.analyzeImpact(files, config, projectRoot),
      });
      const enriched = await enricher.enrich(modeData.path, modeData.content, projectRoot);
      if (enriched !== null) {
        await this.kaos.writeText(modeData.path, enriched);
      }
    } catch {
      // Enrichment is best-effort; failures should not block plan exit.
    }
  }

  private async resolvePlanReviewDisplay(
    args: ExitPlanModeInput,
  ): Promise<ToolInputDisplay | undefined> {
    if (!this.agent.sessionMode.isActive) return undefined;
    let data: SessionModeData;
    try {
      data = await this.agent.sessionMode.data();
    } catch {
      return undefined;
    }
    if (data === null || data.content.trim().length === 0) return undefined;
    const display: ToolInputDisplay = {
      kind: 'plan_review',
      plan: data.content,
      path: data.path,
    };
    if (args.options !== undefined && args.options.length >= 2) {
      display.options = args.options;
    }
    return display;
  }

  private async execution(
    args: ExitPlanModeInput,
    metadata?: unknown,
  ): Promise<ExecutableToolResult> {
    if (!this.agent.sessionMode.isActive) {
      return {
        isError: true,
        output:
          'ExitPlanMode can only be called while plan mode is active. Use EnterPlanMode (or /plan) first.',
      };
    }

    // E2E enrichment: analyze changes and append E2E task after approval but
    // before the plan is handed off, so the file is only modified when the user
    // actually approves the plan.
    await this.maybeEnrichPlanForE2E();

    const resolvedPlan = await this.resolvePlan();
    if (!resolvedPlan.ok) return resolvedPlan.error;

    // When the plan was approved through the review surface, the approval policy
    // already tracked `plan_submitted` (in evaluate()) and forwarded the chosen
    // option via executionMetadata. In auto mode the policy is skipped, so the
    // tool is the only place these events fire.
    const viaApproval = isViaApproval(metadata);
    // The raw approval label (used for telemetry, even reserved labels like "Approve").
    const rawLabel = selectedLabelOf(metadata);
    // The approach prefix is only surfaced for a label that is one of the declared
    // options, so a plain approval ("Approve") never prints "Selected approach: Approve".
    const optionLabel = declaredOptionLabel(args.options, rawLabel);

    if (!viaApproval) {
      this.agent.telemetry.track('plan_submitted', {
        has_options: args.options !== undefined && args.options.length >= 2,
      });
    }

    const failed = await this.handoffToNormal(optionLabel);
    if (failed !== undefined) return failed;

    // Track resolution only AFTER a successful handoff so a failed exit() does not
    // emit a spurious approved/auto_approved event (see exit-plan-mode-telemetry.test.ts).
    if (viaApproval) {
      this.agent.telemetry.track(
        'plan_resolved',
        rawLabel !== undefined && rawLabel.length > 0
          ? { outcome: 'approved', chosen_option: rawLabel }
          : { outcome: 'approved' },
      );
    } else {
      this.agent.telemetry.track('plan_resolved', { outcome: 'auto_approved' });
    }

    return {
      isError: false,
      stopTurn: true,
      output: `Exited plan mode. ${formatPlanHandoffOutput(resolvedPlan.plan, resolvedPlan.path, optionLabel)}`,
    };
  }

  private async handoffToNormal(
    selectedLabel: string | undefined,
  ): Promise<ExecutableToolResult | undefined> {
    try {
      await this.agent.sessionMode.handoffTo('normal', { selectedLabel });
    } catch (error) {
      const message = error instanceof Error ? error.message : 'Failed to exit plan mode.';
      return {
        isError: true,
        output: `Failed to exit plan mode: ${message}`,
      };
    }
  }

  private async resolvePlan(): Promise<ResolvePlanResult> {
    let source: ExitPlanModePlanSource | null;
    try {
      const data = await this.agent.sessionMode.data();
      source = data === null ? null : { plan: data.content, path: data.path };
    } catch (error) {
      const message = error instanceof Error ? error.message : 'Failed to read plan file.';
      return {
        ok: false,
        error: { isError: true, output: `Failed to read plan file: ${message}` },
      };
    }

    if (source !== null && source.plan.trim().length > 0) {
      return {
        ok: true,
        plan: source.plan,
        path: source.path,
      };
    }

    const path = source?.path ?? this.agent.sessionMode.sessionModeFilePath;
    return {
      ok: false,
      error: {
        isError: true,
        output:
          path === null
            ? 'No plan file found. Write the plan to the current plan file first, then call ExitPlanMode.'
            : `No plan file found. Write your plan to ${path} first, then call ExitPlanMode.`,
      },
    };
  }
}

function hasUniqueOptionLabels(options: readonly ExitPlanModeOption[]): boolean {
  const labels = new Set<string>();
  for (const option of options) {
    const label = normalizeOptionLabel(option.label);
    if (labels.has(label)) return false;
    labels.add(label);
  }
  return true;
}

function hasNoReservedOptionLabels(options: readonly ExitPlanModeOption[]): boolean {
  return options.every((option) => !RESERVED_OPTION_LABELS.has(normalizeOptionLabel(option.label)));
}

function normalizeOptionLabel(label: string): string {
  return label.trim().toLowerCase();
}

function formatPlanHandoffOutput(
  plan: string,
  path: string | undefined,
  selectedLabel: string | undefined,
): string {
  const optionPrefix = selectedApproachPrefix(selectedLabel);
  const savedTo = path !== undefined ? `Plan saved to: ${path}\n\n` : '';
  return `${optionPrefix}Plan mode deactivated. The approved plan has been handed off to the main conversation context.\n${savedTo}## Approved Plan:\n${plan}\n\nSTOP — do NOT begin executing now. This turn ends here. The user will start implementation themselves — the plan is now available in their main conversation context.`;
}

File: rust-ody/crates/tools-rs/src/builtin/session_mode/exit_plan_mode.rs
use serde_json::Value;
use std::sync::Arc;

use crate::builtin::session_mode::planning::{
    declared_option_label, is_via_approval, selected_approach_prefix, selected_label_of,
    ExitModeOption,
};
use crate::builtin::session_mode::SessionModeProvider;
use crate::builtin::{
    BuiltinTool, ExecutableToolContext, ExecutableToolOutput, ExecutableToolResult, ToolError,
    ToolExecution,
};

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ExitPlanModeInput {
    #[serde(default)]
    pub options: Vec<ExitModeOption>,
}

pub struct ExitPlanModeTool {
    provider: Arc<dyn SessionModeProvider>,
}

impl ExitPlanModeTool {
    pub fn new(provider: Arc<dyn SessionModeProvider>) -> Self {
        Self { provider }
    }
}

impl BuiltinTool for ExitPlanModeTool {
    fn name(&self) -> &str {
        "ExitPlanMode"
    }

    fn description(&self) -> &str {
        "Present the finalized plan to the user and exit plan mode."
    }

    fn parameters(&self) -> Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "options": {
                    "type": "array",
                    "minItems": 1,
                    "maxItems": 3,
                    "items": {
                        "type": "object",
                        "properties": {
                            "label": { "type": "string", "minLength": 1, "maxLength": 80 },
                            "description": { "type": "string" }
                        },
                        "required": ["label"]
                    },
                    "description": "When the plan contains multiple alternative approaches, list them here so the user can choose which one to execute."
                }
            },
            "additionalProperties": false
        })
    }

    fn resolve_execution(&self, args: Value) -> Result<ToolExecution, ToolError> {
        let input: ExitPlanModeInput =
            serde_json::from_value(args).map_err(|e| ToolError::InvalidArgs(e.to_string()))?;

        let provider = Arc::clone(&self.provider);
        let display = build_plan_review_display(&*provider);

        Ok(ToolExecution {
            accesses: Default::default(),
            description: "Presenting plan and exiting plan mode".into(),
            approval_rule: "ExitPlanMode".into(),
            matches_rule: None,
            display,
            execute: Box::new(move |ctx: ExecutableToolContext| {
                let provider = Arc::clone(&provider);
                let input = input.clone();
                Box::pin(async move { execute_exit_plan_mode(provider, input, ctx).await })
            }),
        })
    }
}

fn build_plan_review_display(provider: &dyn SessionModeProvider) -> Option<Value> {
    if !provider.is_session_mode_active() {
        return None;
    }
    let path = provider.session_mode_file_path()?;
    let content = match read_session_mode_file_sync(provider.kaos().as_ref(), &path) {
        Ok(c) => c,
        Err(_) => return None,
    };
    let trimmed = content.trim();
    if trimmed.is_empty() {
        return None;
    }
    let display = serde_json::json!({
        "kind": "plan_review",
        "plan": trimmed,
        "path": path,
    });
    Some(display)
}

fn read_session_mode_file_sync(
    kaos: &dyn crate::builtin::session_mode::SessionModeContext,
    path: &str,
) -> anyhow::Result<String> {
    tokio::task::block_in_place(|| tokio::runtime::Handle::current().block_on(kaos.read_text(path)))
}

async fn execute_exit_plan_mode(
    provider: Arc<dyn SessionModeProvider>,
    input: ExitPlanModeInput,
    ctx: ExecutableToolContext,
) -> ExecutableToolResult {
    if !provider.is_session_mode_active() {
        return ExecutableToolResult::error_text(
            "ExitPlanMode can only be called while plan mode is active. Use EnterPlanMode (or /plan) first.".into(),
            "not in plan mode".into(),
        );
    }

    let (plan, path) = match resolve_plan(&*provider).await {
        Ok(v) => v,
        Err(e) => return e,
    };

    let option_label = declared_option_label(
        Some(&input.options),
        selected_label_of(ctx.metadata.as_ref()).as_deref(),
    );

    let metadata = ctx.metadata.as_ref();
    if !is_via_approval(metadata) {
        provider.telemetry().track(
            "plan_submitted",
            std::collections::HashMap::from([(
                "has_options".into(),
                serde_json::Value::Bool(input.options.len() >= 2),
            )]),
        );
    }

    if let Err(e) = provider.handoff_to("normal", option_label.clone()).await {
        return ExecutableToolResult::error_text(
            format!("Failed to exit plan mode: {}", e),
            "handoff failed".into(),
        );
    }

    if is_via_approval(metadata) {
        let raw_label = selected_label_of(metadata);
        let props = if let Some(l) = raw_label {
            std::collections::HashMap::from([
                ("outcome".into(), "approved".into()),
                ("chosen_option".into(), l.into()),
            ])
        } else {
            std::collections::HashMap::from([("outcome".into(), "approved".into())])
        };
        provider.telemetry().track("plan_resolved", props);
    } else {
        provider.telemetry().track(
            "plan_resolved",
            std::collections::HashMap::from([("outcome".into(), "auto_approved".into())]),
        );
    }

    let path_line = path
        .as_ref()
        .map(|p| format!("Plan saved to: {}\n\n", p))
        .unwrap_or_default();

    let output = format!(
        "{}Exited plan mode. {}Plan mode deactivated. The approved plan has been handed off to the main conversation context.\n\n## Approved Plan:\n{}\n\nSTOP — do NOT begin executing now. This turn ends here. The user will start implementation themselves — the plan is now available in their main conversation context.",
        selected_approach_prefix(option_label.as_deref()),
        path_line,
        plan
    );

    ExecutableToolResult {
        output: ExecutableToolOutput::Text(output),
        message: None,
        is_error: false,
        stop_turn: Some(true),
    }
}

async fn resolve_plan(
    provider: &dyn SessionModeProvider,
) -> Result<(String, Option<String>), ExecutableToolResult> {
    let path = provider.session_mode_file_path();
    let content = match path.as_ref() {
        Some(p) => match provider.kaos().read_text(p).await {
            Ok(c) => c,
            Err(e) => {
                return Err(ExecutableToolResult::error_text(
                    format!("Failed to read plan file: {}", e),
                    "read failed".into(),
                ));
            }
        },
        None => String::new(),
    };

    if content.trim().is_empty() {
        return Err(ExecutableToolResult::error_text(
            match path {
                Some(p) => format!("No plan file found. Write your plan to {} first, then call ExitPlanMode.", p),
                None => "No plan file found. Write the plan to the current plan file first, then call ExitPlanMode.".into(),
            },
            "empty plan".into(),
        ));
    }

    Ok((content, path))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::builtin::session_mode::tests::MockSessionModeProvider;

    #[tokio::test(flavor = "multi_thread")]
    async fn exit_plan_mode_hands_off_to_normal() {
        let provider = Arc::new(MockSessionModeProvider::plan_mode_with_content(
            "## Plan\n\nDo X.",
        ));
        let tool = ExitPlanModeTool::new(provider.clone());
        let exec = tool.resolve_execution(serde_json::json!({})).unwrap();
        assert!(exec.display.is_some());
        let result = (exec.execute)(ExecutableToolContext {
            turn_id: "1".into(),
            tool_call_id: "call_1".into(),
            signal: crate::builtin::AbortSignal::new(),
            metadata: None,
        })
        .await;
        assert!(!result.is_error);
        assert_eq!(result.stop_turn, Some(true));
        let text = result.to_text();
        assert!(text.contains("Plan mode deactivated"));
        assert!(provider
            .handed_off_to
            .lock()
            .unwrap()
            .contains(&("normal".to_string(), None)));
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn exit_plan_mode_preserves_selected_label() {
        let provider = Arc::new(MockSessionModeProvider::plan_mode_with_content(
            "## Plan\n\nDo X.",
        ));
        let tool = ExitPlanModeTool::new(provider.clone());
        let exec = tool
            .resolve_execution(serde_json::json!({
                "options": [{"label": "Fast", "description": ""}]
            }))
            .unwrap();
        let result = (exec.execute)(ExecutableToolContext {
            turn_id: "1".into(),
            tool_call_id: "call_1".into(),
            signal: crate::builtin::AbortSignal::new(),
            metadata: None,
        })
        .await;
        assert!(!result.is_error);
        assert!(provider
            .handed_off_to
            .lock()
            .unwrap()
            .contains(&("normal".to_string(), None)));
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn exit_plan_mode_errors_when_inactive() {
        let provider = Arc::new(MockSessionModeProvider::inactive());
        let tool = ExitPlanModeTool::new(provider.clone());
        let exec = tool.resolve_execution(serde_json::json!({})).unwrap();
        let result = (exec.execute)(ExecutableToolContext {
            turn_id: "1".into(),
            tool_call_id: "call_1".into(),
            signal: crate::builtin::AbortSignal::new(),
            metadata: None,
        })
        .await;
        assert!(result.is_error);
        assert!(result
            .to_text()
            .contains("ExitPlanMode can only be called while plan mode is active"));
    }
}
